File: y2019-rust/src/day01.rs
use std::cmp::max;
use std::fs::File;
use std::io::{BufRead, BufReader};

pub fn solve() {
    println!("{}\n{}", solve1(), solve2());
}

fn solve1() -> i32 {
    let xs = read();
    let mut res = 0;
    for x in xs {
        res += x / 3 - 2;
    }

    res
}

fn solve2() -> i32 {
    let xs = read();
    let mut res = 0;
    for mut x in xs {
        while x > 0 {
            x = max(x / 3 - 2, 0);
            res += x;
        }
    }

    res
}

fn read() -> Vec<i32> {
    let file = File::open("files/01.in").expect("No file");
    let reader = BufReader::new(file);

    reader.lines()
        .map(|l| l.map(|s| s.parse::<i32>().expect("Could not parse")).expect(""))
        .collect()
}

File: y2020-julia/src/day01.jl
function read(filename)
    data = Int[]

    f = open(filename, "r")
    for ln in eachline(f)
        push!(data, parse(Int, ln))
    end
    close(f)

    return data
end

function solve1(input)
    n = length(input)
    for i in 1:n
        for j in (i + 1):n
            if input[i] + input[j] == 2020
                return input[i] * input[j]
            end
        end
    end
end

function solve2(input)
    n = length(input)
    for i in 1:n
        for j in (i + 1):n
            for k in (j + 1):n
                if input[i] + input[j] + input[k] == 2020
                    return input[i] * input[j] * input[k]
                end
            end
        end
    end
end

function main()
    input = read("files/01.in")
    @time println(solve1(input))
    @time println(solve2(input))
end

main()